server/src/main/java/org/opensearch/cluster/decommission/DecommissionController.java

```diff
@@ -267,4 +277,128 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
+        assert decommissionAttributeMetadata.status().equals(DecommissionStatus.WEIGH_AWAY)
+            : "unexpected status encountered while decommissioning nodes";
+        DecommissionAttribute decommissionAttribute = decommissionAttributeMetadata.decommissionAttribute();
+
+        Map<String, String> weights = new HashMap<>();
+        zones.forEach(zone -> {
+            if (zone.equalsIgnoreCase(decommissionAttribute.attributeValue())) {
+                weights.put(zone, "0");
+            } else {
+                weights.put(zone, "1");
+            }
+        });
+
+        // WRR API will validate invalid weights
+        final ClusterPutWRRWeightsRequest clusterWeightRequest = new ClusterPutWRRWeightsRequest();
+        clusterWeightRequest.attributeName(decommissionAttribute.attributeName());
+        clusterWeightRequest.setWRRWeight(weights);
+
+        transportService.sendRequest(
+            transportService.getLocalNode(),
+            ClusterPutWRRWeightsAction.NAME,
+            clusterWeightRequest,
+            new TransportResponseHandler<ClusterPutWRRWeightsResponse>() {
+                @Override
+                public void handleResponse(ClusterPutWRRWeightsResponse response) {
+                    logger.info("Weights are successfully set.");
+                    listener.onResponse(response);
+                }
+
+                @Override
+                public void handleException(TransportException exp) {
+                    // Logging warn message on failure. Should we do Retry? If weights are not set should we fail?
+                    logger.warn("Exception occurred while setting weights.Exception Messages - [{}]", exp.unwrapCause().getMessage());
+                    listener.onFailure(exp);
+                }
+
+                @Override
+                public String executor() {
+                    return ThreadPool.Names.SAME;
+                }
+
+                @Override
+                public ClusterPutWRRWeightsResponse read(StreamInput in) throws IOException {
+                    return new ClusterPutWRRWeightsResponse(in);
+                }
+            }
+        );
+    }
+
+    void handleNodesDecommissionRequest(
+        Set<DiscoveryNode> decommissionedNodes,
+        String reason,
+        TimeValue timeout,
+        TimeValue timeoutForNodeDraining,
+        ActionListener<Void> listener
+    ) {
+        // Wait for timeout to happen. Log the active connection before decommissioning of nodes.
+        transportService.getThreadPool().schedule(new Runnable() {
+            @Override
+            public void run() {
+                // Check for active connections.
+                getActiveRequestCountOnDecommissionNodes(decommissionedNodes);
+                removeDecommissionedNodes(decommissionedNodes, reason, timeout, listener);
+            }
+        }, timeoutForNodeDraining, org.opensearch.threadpool.ThreadPool.Names.SAME);
+    }
```

Comment:
Didn't get this. `removeDecommissionedNodes` should only be called once after node is drained